### DISCUSSION TITLE
add sf polygon data

### DISCUSSION THREAD

**Discussion by wiesehahn** [r-lidar/lidR#677] [General]
in lidR there are the specific functions `add_dtm3d`, `add_treetops3d`, `add_flightlines3d`. 
Although they sound quite specific I guess they could be seen more or less as `add_raster3d`, `add_sf_points3d`, `add_sf_lines3d` and used to add raster, points and lines in general?

Is there any way to add polygon data? (e.g. I want to overlay forest disturbance data from sf polygons)
I guess maybe directly via rgl.

**Comment by wiesehahn** [comment]
sorry for opening this discussion, under https://r-lidar.github.io/lidRbook/io.html#plot there is a hint towards rgl functions, I guess `rgl::polygon3d` will do

**Reply by Jean-Romain** [reply]
The challenge is that you need a 3D polygons with a Z coordinate otherwise the default will be to print at Z = 0 which may not suit for the plot. By the way please open an issue with a clear statement of what you would like to see as plot functions with examples and so on.

**Reply by wiesehahn** [reply]
I can definitely state some things in the next time I would like to see if that is wanted. Just didn't open an issue because it is not a bug and also didn't want to annoy you with feature requests (unfortunately my skills are not good enough to add qualified pull requests)

In this case I wanted to create a stacked view and plot the polygon above the other data so this did the trick:

```R
offsets <- plot(las)

x <- sf::st_coordinates(polygon)[,1] - offsets[1] 
y <- sf::st_coordinates(polygon)[,2] - offsets[2] 
z <- max(las$Z) + 100
M <- cbind(x,y,z)


rgl::polygon3d(M)
```

![stack_dtm_trees_disturance](https://user-images.githubusercontent.com/41429613/235129848-f674425a-8108-4710-ac4a-ae3654bbf8bc.png)

**Reply by Jean-Romain** [reply]
Nice ! This is definitively the kind of display I wanted people to be able to make when I designed lidR.

Issues are for bug and feature requests. They allow me to keep track of what is opened and what is closed and what I have to do. This fits perfectly in a issue and you already have a code snippet to show me. So please open a detailed issue with code, images, example of use cases. This will help me a lot if you provide me a tidy plan with your ideas. Thank you.

**Reply by wiesehahn** [reply]
Thanks, I will see what I can do next week

**Reply by candelas762** [reply]
I am quite impress on how yor image came up @wiesehahn . Would it be possible to provide some code to replicate it?

**Reply by wiesehahn** [reply]
hey probably better late than never, you can try something like this ( I didnt make this reproducible here):

```r
#load point cloud
las <- readLAS(LASfile)

# merge disturbance data
disturbance <- st_read(here("data/external/disturbance_polygon.geojson"))
disturbance <- st_transform(disturbance, st_crs(las))
las <- merge_spatial(las, disturbance, attribute = "disturbance")


# calculate normalized height and add as attribute
dtm <- grid_terrain(las, res = 0.5, algorithm = knnidw(k = 10L, p = 2), full_raster =T)

nlas <- normalize_height(las, dtm, add_lasattribute=TRUE)

las <- add_lasattribute(las, nlas@data$Z, "height", "normalized point height")

# plot disturbed and undisturbed
nondist <- filter_poi(las, disturbance == FALSE & Classification !=2)
dist <- filter_poi(las, disturbance == TRUE & Classification !=2)

nondist$Z <- nondist$Z+100
dist$Z <- dist$Z+100

palette_col <- viridis::plasma(100)
palette_desat <- colorspace::lighten(palette_col, 0.7)
palette_desat <- colorspace::desaturate(palette_desat, 0.5)



disturbance$z <- max(nondist$Z) + 100
disturbance <- st_sf(disturbance)


ttops <- locate_trees(dist, lmf(ws = 7, hmin = 10))
ttops$Z <- ttops$Z +50
ttops <- st_zm(ttops, drop=TRUE)





offsets <- plot(decimate_points(nondist, random(4)), color = "height", bg = "white", pal = palette_desat, size = 1)

# disturbance polygon for rgl
x <- sf::st_coordinates(disturbance)[,1] - offsets[1] 
y <- sf::st_coordinates(disturbance)[,2] - offsets[2] 
z <- max(nondist$Z) + 100
M <- cbind(x,y,z)

plot(decimate_points(dist, random(4)), color = "height", bg = "white", pal = palette_col, size = 1, add = offsets) |> 
  add_dtm3d(dtm) |>  
  add_treetops3d(ttops, z="Z", color = "black")

rgl::polygon3d(M, fill = FALSE)


# get view selected interactively
um <- rgl::par3d()$userMatrix


# set view
rgl::view3d(userMatrix = um)

rgl::snapshot3d("name.png")

```
